File: kyonggiyo-core/src/main/java/kyonggiyo/application/service/candidate/CandidateQueryService.java
package kyonggiyo.application.service.candidate;

import kyonggiyo.application.port.in.auth.dto.UserInfo;
import kyonggiyo.application.port.in.candidate.LoadCandidateUseCase;
import kyonggiyo.application.port.in.candidate.dto.CandidateResponse;
import kyonggiyo.application.port.out.candidate.LoadCandidatePort;
import kyonggiyo.domain.candidate.Candidate;
import kyonggiyo.domain.candidate.Status;
import kyonggiyo.common.response.SliceResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Slice;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class CandidateQueryService implements LoadCandidateUseCase {

    private static final int DEFAULT_PAGE_SIZE = 10;

    private final LoadCandidatePort loadCandidatePort;

    @Override
    public SliceResponse<CandidateResponse> findAllByStatus(UserInfo userInfo, Status status, int page) {
        Pageable pageable = PageRequest.of(page, DEFAULT_PAGE_SIZE);
        Slice<Candidate> candidates = loadCandidatePort.findAllByStatus(status, pageable);
        return SliceResponse.of(candidates.map(CandidateResponse::from));
    }

}

File: kyonggiyo-core/src/main/java/kyonggiyo/application/candidate/service/CandidateQueryService.kt
package kyonggiyo.application.candidate.service

import kyonggiyo.application.auth.domain.vo.UserInfo
import kyonggiyo.application.candidate.domain.vo.Status
import kyonggiyo.application.candidate.port.inbound.CandidateResponse
import kyonggiyo.application.candidate.port.inbound.CandidateResponse.Companion.from
import kyonggiyo.application.candidate.port.inbound.LoadCandidateUseCase
import kyonggiyo.application.candidate.port.outbound.LoadCandidatePort
import kyonggiyo.common.response.SliceResponse
import org.springframework.data.domain.PageRequest
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class CandidateQueryService(
    private val loadCandidatePort: LoadCandidatePort,
) : LoadCandidateUseCase {

    override fun findAllByStatus(userInfo: UserInfo, status: Status, page: Int): SliceResponse<CandidateResponse> {
        return PageRequest.of(page, DEFAULT_PAGE_SIZE)
            .let { loadCandidatePort.findAllByStatus(status, it) }
            .let { SliceResponse.of(it.map { response -> from(response) }) }
    }

    companion object {
        private const val DEFAULT_PAGE_SIZE = 10
    }

}
